// luma.gl/dist/shadertools/shaders/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MODULAR_SHADERS = exports.MONOLITHIC_SHADERS_2 = exports.MONOLITHIC_SHADERS = exports.SHADERS = undefined;

var _shaderlib = require('../../deprecated/shaderlib');

Object.defineProperty(exports, 'SHADERS', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_shaderlib).default;
  }
});
Object.defineProperty(exports, 'MONOLITHIC_SHADERS', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_shaderlib).default;
  }
});

var _defaultVertex = require('./default-vertex.glsl');

var _defaultVertex2 = _interopRequireDefault(_defaultVertex);

var _defaultFragment = require('./default-fragment.glsl');

var _defaultFragment2 = _interopRequireDefault(_defaultFragment);

var _modularVertex = require('./modular-vertex.glsl');

var _modularVertex2 = _interopRequireDefault(_modularVertex);

var _modularFragment = require('./modular-fragment.glsl');

var _modularFragment2 = _interopRequireDefault(_modularFragment);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// A set of monolithic shaders


var defaultUniforms = require('./default-uniforms');

var MONOLITHIC_SHADERS_2 = exports.MONOLITHIC_SHADERS_2 = {
  vs: _defaultVertex2.default,
  fs: _defaultFragment2.default,
  defaultUniforms: defaultUniforms
};

// A set of base shaders that leverage the shader module system,
// dynamically enabling features depending on which modules are included
var MODULAR_SHADERS = exports.MODULAR_SHADERS = {
  vs: _modularVertex2.default,
  fs: _modularFragment2.default,
  defaultUniforms: {}
};
