// luma.gl/dist/packages/math/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _vector = require('./src/vector2');

Object.defineProperty(exports, 'Vector2', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_vector).default;
  }
});

var _vector2 = require('./src/vector3');

Object.defineProperty(exports, 'Vector3', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_vector2).default;
  }
});

var _vector3 = require('./src/vector4');

Object.defineProperty(exports, 'Vector4', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_vector3).default;
  }
});

var _matrix = require('./src/matrix4');

Object.defineProperty(exports, 'Matrix4', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_matrix).default;
  }
});

var _quaternion = require('./src/quaternion');

Object.defineProperty(exports, 'Quaternion', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_quaternion).default;
  }
});

var _common = require('./src/common');

Object.defineProperty(exports, 'config', {
  enumerable: true,
  get: function get() {
    return _common.config;
  }
});
Object.defineProperty(exports, 'checkNumber', {
  enumerable: true,
  get: function get() {
    return _common.checkNumber;
  }
});
Object.defineProperty(exports, 'configure', {
  enumerable: true,
  get: function get() {
    return _common.configure;
  }
});
Object.defineProperty(exports, 'formatValue', {
  enumerable: true,
  get: function get() {
    return _common.formatValue;
  }
});
Object.defineProperty(exports, 'isArray', {
  enumerable: true,
  get: function get() {
    return _common.isArray;
  }
});
Object.defineProperty(exports, 'clone', {
  enumerable: true,
  get: function get() {
    return _common.clone;
  }
});
Object.defineProperty(exports, 'radians', {
  enumerable: true,
  get: function get() {
    return _common.radians;
  }
});
Object.defineProperty(exports, 'degrees', {
  enumerable: true,
  get: function get() {
    return _common.degrees;
  }
});
Object.defineProperty(exports, 'sin', {
  enumerable: true,
  get: function get() {
    return _common.sin;
  }
});
Object.defineProperty(exports, 'cos', {
  enumerable: true,
  get: function get() {
    return _common.cos;
  }
});
Object.defineProperty(exports, 'tan', {
  enumerable: true,
  get: function get() {
    return _common.tan;
  }
});
Object.defineProperty(exports, 'asin', {
  enumerable: true,
  get: function get() {
    return _common.asin;
  }
});
Object.defineProperty(exports, 'acos', {
  enumerable: true,
  get: function get() {
    return _common.acos;
  }
});
Object.defineProperty(exports, 'atan', {
  enumerable: true,
  get: function get() {
    return _common.atan;
  }
});
Object.defineProperty(exports, 'clamp', {
  enumerable: true,
  get: function get() {
    return _common.clamp;
  }
});
Object.defineProperty(exports, 'equals', {
  enumerable: true,
  get: function get() {
    return _common.equals;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
